sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala

```diff
@@ -2851,6 +2851,68 @@ class AdaptiveQueryExecSuite
     val unionDF = aggDf1.union(aggDf2)
     checkAnswer(unionDF.select("id").distinct, Seq(Row(null)))
   }
+
+  test("SPARK-44065: Optimize BroadcastHashJoin skew when localShuffleReader is disabled") {
+    withSQLConf(
+      SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true",
+      SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1",
+      SQLConf.ADAPTIVE_AUTO_BROADCASTJOIN_THRESHOLD.key -> "1000",
+      SQLConf.LOCAL_SHUFFLE_READER_ENABLED.key -> "false",
+      SQLConf.SHUFFLE_PARTITIONS.key -> "10",
+      SQLConf.SKEW_JOIN_SKEWED_PARTITION_THRESHOLD.key -> "600",
+      SQLConf.ADVISORY_PARTITION_SIZE_IN_BYTES.key -> "600") {
+      withTempView("skewData", "smallData") {
+        spark
+          .range(0, 1000, 1, 10)
+          .selectExpr("if(id >= 5, 5, id) as key1", "id as value1")
+          .createOrReplaceTempView("skewData1")
+        spark
+          .range(0, 5, 1, 10)
+          .selectExpr("id key2", "id as value2")
+          .createOrReplaceTempView("smallData")
+
+        Seq(true, false).foreach { localShuffleReader =>
+          withSQLConf(SQLConf.LOCAL_SHUFFLE_READER_ENABLED.key -> localShuffleReader.toString) {
+            val sqlText =
+              s"""
+                 |select * from skewData1 a join smallData b on a.key1 = b.key2
+                 |""".stripMargin
+            val (_, plan) = runAdaptiveAndVerifyResult(sqlText)
+            val bhjs = findTopLevelBroadcastHashJoin(plan)
+            assert(bhjs.nonEmpty)
+
+            if (localShuffleReader) {
+              val localShuffleReaders = collect(plan) {
+                case c: AQEShuffleReadExec if c.isLocalRead => c
```

Comment:
I think there is a issue with local shuffle read. Before, the rule `OptimizeShuffleWithLocalRead#getPartitionSpecs` assume all the shuffle block are the same data size, https://github.com/apache/spark/blob/29e09473a816f1cb0ddea3e49d6dcb492ad473d9/sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeShuffleWithLocalRead.scala#L67-L69
but now the assumption is broken. We should update the code to consider skewed partitions, otherwise it can be still skewed when enabled local shuffle read.

Reply:
it seems not addressed, can you take a look ? @wForget

Reply:
> it seems not addressed, can you take a look ? @wForget

Sorry, I'm still confused about this so belatedly did it. Do we still need to follow advisoryParallelism if we consider block size?

Reply:
We should respect `PartialReducerPartitionSpec` in `OptimizeShuffleWithLocalRead`. Before this pr, `OptimizeShuffleWithLocalRead` assumes it would not accpet the skewed partition, so we should make it consider `PartialReducerPartitionSpec`.

Reply:
Do you mean we don't need `OptimizeShuffleWithLocalRead` optimization when there are `PartialReducerPartitionSpec`?

Reply:
I have skipped optimization in `OptimizeShuffleWithLocalRead` for `AQEShuffleReadExec` with skewed partitions. What do you think of this?